executor/join.go

```diff
@@ -208,9 +261,128 @@ func (e *HashJoinExec) fetchOuterRows(goCtx goctx.Context) {
 	}
 }
 
-// prepare runs the first time when 'Next' is called, it starts one worker goroutine to fetch rows from the big table,
-// and reads all data from the small table to build a hash table, then starts multiple join worker goroutines.
-func (e *HashJoinExec) prepare(goCtx goctx.Context) error {
+// fetchOuterChunks get chunks from fetches chunks from the big table in a background goroutine
+// and sends the chunks to multiple channels which will be read by multiple join workers.
+func (e *HashJoinExec) fetchOuterChunks(goCtx goctx.Context) {
+	defer func() {
+		for i := range e.outerResultChs {
+			close(e.outerResultChs[i])
+		}
+		e.workerWaitGroup.Done()
+	}()
+	for i := 0; ; i = (i + 1) % e.concurrency {
+		if e.finished.Load().(bool) {
+			return
+		}
+		outerResult := &execWorkerResult{
+			src: e.outerChkResource,
+		}
+
+		select {
+		case _, ok := <-e.closeCh:
+			if !ok {
+				return
+			}
+		case outerResult.chk = <-e.outerChkResource:
```

Comment:
We can make `e.outerChkResource` be a `chan` of `execWorkerResult`, once we received a `execWorkerResult`, we fetch the outer chunk and put that chunk into `execWorkerResult.src`.

This is more precise to decide which channel we should put the outer chunk into.